File: libs/language/src/lib/model/Node.ts
import {
  isNonNullable,
  isNullable,
  JsonPrimitive,
} from '@blue-labs/shared-utils';
import { NodePathAccessor } from '../utils/NodePathAccessor';
import { BigIntegerNumber } from './BigIntegerNumber';
import { BigDecimalNumber } from './BigDecimalNumber';
import {
  BOOLEAN_TYPE_BLUE_ID,
  DOUBLE_TYPE_BLUE_ID,
  INTEGER_TYPE_BLUE_ID,
} from '../utils/Properties';

export class BlueNode {
  static INTEGER: BlueNode = new BlueNode('Integer');

  private name?: string;
  private description?: string;
  private type?: BlueNode;
  private itemType?: BlueNode;
  private keyType?: BlueNode;
  private valueType?: BlueNode;
  private value?:
    | Exclude<JsonPrimitive, number>
    | BigIntegerNumber
    | BigDecimalNumber;
  private items?: BlueNode[];
  private properties?: Record<string, BlueNode>;
  private blueId?: string;
  private blue?: BlueNode;
  private inlineValue = false;

  constructor(name?: string) {
    this.name = name;
  }

  getName() {
    return this.name;
  }

  setName(name: string | undefined): BlueNode {
    this.name = name;
    return this;
  }

  getDescription() {
    return this.description;
  }

  setDescription(description: string | undefined): BlueNode {
    this.description = description;
    return this;
  }

  getType() {
    return this.type;
  }

  setType(type: BlueNode | string | undefined): BlueNode {
    if (typeof type === 'string') {
      this.type = new BlueNode().setValue(type).setInlineValue(true);
    } else {
      this.type = type;
    }
    return this;
  }

  getItemType() {
    return this.itemType;
  }

  setItemType(itemType: BlueNode | string | undefined): BlueNode {
    if (typeof itemType === 'string') {
      this.itemType = new BlueNode().setValue(itemType).setInlineValue(true);
    } else {
      this.itemType = itemType;
    }
    return this;
  }

  getKeyType() {
    return this.keyType;
  }

  setKeyType(keyType: BlueNode | string | undefined): BlueNode {
    if (typeof keyType === 'string') {
      this.keyType = new BlueNode().setValue(keyType).setInlineValue(true);
    } else {
      this.keyType = keyType;
    }
    return this;
  }

  getValueType() {
    return this.valueType;
  }

  setValueType(valueType: BlueNode | string | undefined): BlueNode {
    if (typeof valueType === 'string') {
      this.valueType = new BlueNode().setValue(valueType).setInlineValue(true);
    } else {
      this.valueType = valueType;
    }
    return this;
  }

  getValue() {
    const typeBlueId = this.type?.getBlueId();
    if (isNonNullable(typeBlueId) && isNonNullable(this.value)) {
      if (
        typeBlueId === INTEGER_TYPE_BLUE_ID &&
        typeof this.value === 'string'
      ) {
        return new BigIntegerNumber(this.value);
      } else if (
        typeBlueId === DOUBLE_TYPE_BLUE_ID &&
        typeof this.value === 'string'
      ) {
        const parsed = new BigDecimalNumber(this.value);
        const doubleValue = parseFloat(parsed.toString());
        return new BigDecimalNumber(doubleValue.toString());
      } else if (
        typeBlueId === BOOLEAN_TYPE_BLUE_ID &&
        typeof this.value === 'string'
      ) {
        return this.value.toLowerCase() === 'true';
      }
    }
    return this.value;
  }

  setValue(
    value: JsonPrimitive | BigIntegerNumber | BigDecimalNumber | number
  ): BlueNode {
    if (typeof value === 'number') {
      if (value % 1 === 0) {
        this.value = new BigIntegerNumber(value.toString());
      } else {
        this.value = new BigDecimalNumber(value.toString());
      }
    } else {
      this.value = value;
    }
    return this;
  }

  getItems() {
    return this.items;
  }

  setItems(items: BlueNode[] | undefined): BlueNode {
    this.items = items;
    return this;
  }

  addItems(...items: BlueNode[]): BlueNode {
    if (!this.items) {
      this.items = [];
    }
    this.items.push(...items);
    return this;
  }

  getProperties() {
    return this.properties;
  }

  setProperties(properties: Record<string, BlueNode> | undefined): BlueNode {
    this.properties = properties;
    return this;
  }

  addProperty(key: string, value: BlueNode): BlueNode {
    if (!this.properties) {
      this.properties = {};
    }
    this.properties[key] = value;
    return this;
  }

  removeProperty(key: string): BlueNode {
    if (this.properties) {
      delete this.properties[key];
    }
    return this;
  }

  private getContractsProperty() {
    return this.properties?.['contracts'];
  }

  getContracts() {
    return this.getContractsProperty()?.getProperties();
  }

  setContracts(contracts: Record<string, BlueNode> | undefined): BlueNode {
    if (isNullable(contracts)) {
      this.removeProperty('contracts');
    } else {
      this.addProperty('contracts', new BlueNode().setProperties(contracts));
    }
    return this;
  }

  addContract(key: string, value: BlueNode): BlueNode {
    const contractsProperty = this.getContractsProperty();
    if (isNullable(contractsProperty)) {
      this.addProperty('contracts', new BlueNode().addProperty(key, value));
    } else {
      contractsProperty.addProperty(key, value);
    }
    return this;
  }

  removeContract(key: string): BlueNode {
    const contractsProperty = this.getContractsProperty();
    if (contractsProperty) {
      contractsProperty.removeProperty(key);

      const remainingProperties = contractsProperty.getProperties();
      if (
        remainingProperties &&
        Object.keys(remainingProperties).length === 0
      ) {
        this.removeProperty('contracts');
      }
    }
    return this;
  }

  getBlueId() {
    return this.blueId;
  }

  setBlueId(blueId: string | undefined): BlueNode {
    this.blueId = blueId;
    return this;
  }

  getBlue() {
    return this.blue;
  }

  setBlue(blue: BlueNode | undefined): BlueNode {
    this.blue = blue;
    return this;
  }

  isInlineValue() {
    return this.inlineValue;
  }

  setInlineValue(inlineValue: boolean): BlueNode {
    this.inlineValue = inlineValue;
    return this;
  }

  get(
    path: string,
    linkingProvider?: (node: BlueNode) => BlueNode | null
  ):
    | BlueNode
    | string
    | boolean
    | BigIntegerNumber
    | BigDecimalNumber
    | null
    | undefined {
    return NodePathAccessor.get(this, path, linkingProvider);
  }

  getAsNode(path: string): BlueNode | undefined {
    const value = this.get(path);
    if (value instanceof BlueNode) {
      return value;
    }

    throw new Error(`Value at path ${path} is not a BlueNode: ${value}`);
  }

  getAsInteger(path: string): number | undefined {
    const value = this.get(path);
    if (
      value instanceof BigIntegerNumber ||
      value instanceof BigDecimalNumber
    ) {
      return value.toNumber();
    }

    throw new Error(
      `Value at path ${path} is not a BigInteger or BigDecimal: ${value}`
    );
  }

  clone(): BlueNode {
    const cloned = new BlueNode(this.name);
    cloned.description = this.description;
    cloned.type = this.type?.clone();
    cloned.itemType = this.itemType?.clone();
    cloned.keyType = this.keyType?.clone();
    cloned.valueType = this.valueType?.clone();
    cloned.value = this.value;
    cloned.items = this.items?.map((item) => item.clone());
    if (this.properties) {
      cloned.properties = Object.fromEntries(
        Object.entries(this.properties).map(([k, v]) => [k, v.clone()])
      );
    }
    cloned.blueId = this.blueId;
    cloned.blue = this.blue?.clone();
    cloned.inlineValue = this.inlineValue;
    return cloned;
  }

  toString(): string {
    return `BlueNode{name='${this.name}', description='${this.description}', type=${this.type}, itemType=${this.itemType}, keyType=${this.keyType}, valueType=${this.valueType}, value=${this.value}, items=${this.items}, properties=${this.properties}, blueId='${this.blueId}', blue=${this.blue}, inlineValue=${this.inlineValue}}`;
  }
}

File: references/java-sdk/src/main/java/blue/language/model/Node.java
package blue.language.model;

import blue.language.utils.NodePathAccessor;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static blue.language.utils.Properties.*;

@JsonDeserialize(using = NodeDeserializer.class)
@JsonSerialize(using = NodeSerializer.class)
public class Node implements Cloneable {

    private String name;
    private String description;
    private Node type;
    private Node itemType;
    private Node keyType;
    private Node valueType;
    private Object value;
    private List<Node> items;
    private Map<String, Node> properties;
    private String blueId;
    private transient String computedBlueId;
    private Constraints constraints;
    private Node blue;
    private boolean inlineValue;

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public Node getType() {
        return type;
    }

    public Node getItemType() {
        return itemType;
    }
    
    public Node getKeyType() {
        return keyType;
    }

    public Node getValueType() {
        return valueType;
    }

    public Object getValue() {
        if (this.type != null && this.type.getBlueId() != null && this.value != null) {
            String typeBlueId = this.type.getBlueId();
            if (INTEGER_TYPE_BLUE_ID.equals(typeBlueId) && this.value instanceof String) {
                return new BigInteger((String) this.value);
            } else if (DOUBLE_TYPE_BLUE_ID.equals(typeBlueId) && this.value instanceof String) {
                BigDecimal parsed = new BigDecimal((String) this.value);
                double doubleValue = parsed.doubleValue();
                return new BigDecimal(Double.toString(doubleValue));
            } else if (BOOLEAN_TYPE_BLUE_ID.equals(typeBlueId) && this.value instanceof String) {
                return Boolean.parseBoolean((String) this.value);
            }
        }
        return value;
    }

    public List<Node> getItems() {
        return items;
    }

    public Map<String, Node> getProperties() {
        return properties;
    }

    public String getBlueId() {
        return blueId;
    }

    public String getReferenceBlueId() {
        return blueId;
    }

    public void setReferenceBlueId(String referenceBlueId) {
        this.blueId = referenceBlueId;
    }

    public String getComputedBlueId() {
        return computedBlueId;
    }

    public void setComputedBlueId(String computedBlueId) {
        this.computedBlueId = computedBlueId;
    }

    public Constraints getConstraints() {
        return constraints;
    }

    public Node getBlue() {
        return blue;
    }
    
    public boolean isInlineValue() {
        return inlineValue;
    }

    public Node name(String name) {
        this.name = name;
        return this;
    }

    public Node description(String description) {
        this.description = description;
        return this;
    }

    public Node type(Node type) {
        this.type = type;
        return this;
    }

    public Node type(String type) {
        this.type = new Node().value(type).inlineValue(true);
        return this;
    }

    public Node itemType(Node itemType) {
        this.itemType = itemType;
        return this;
    }

    public Node itemType(String itemType) {
        this.itemType = new Node().value(itemType).inlineValue(true);
        return this;
    }

    public Node keyType(Node keyType) {
        this.keyType = keyType;
        return this;
    }

    public Node keyType(String keyType) {
        this.keyType = new Node().value(keyType).inlineValue(true);
        return this;
    }

    public Node valueType(Node valueType) {
        this.valueType = valueType;
        return this;
    }

    public Node valueType(String valueType) {
        this.valueType = new Node().value(valueType).inlineValue(true);
        return this;
    }

    public Node value(Object value) {
        if (value instanceof BigInteger || value instanceof BigDecimal) {
            this.value = value;
        } else if (value instanceof Float || value instanceof Double) {
            double decimalValue = ((Number) value).doubleValue();
            if (!Double.isFinite(decimalValue)) {
                throw new IllegalArgumentException("Non-finite numeric value is not supported: " + value);
            }
            this.value = BigDecimal.valueOf(decimalValue);
        } else if (value instanceof Number) {
            this.value = BigInteger.valueOf(((Number) value).longValue());
        } else {
            this.value = value;
        }
        return this;
    }

    public Node value(long value) {
        this.value = BigInteger.valueOf(value);
        return this;
    }

    public Node value(double value) {
        if (!Double.isFinite(value)) {
            throw new IllegalArgumentException("Non-finite numeric value is not supported: " + value);
        }
        this.value = BigDecimal.valueOf(value);
        return this;
    }

    public Node items(List<Node> items) {
        this.items = items;
        return this;
    }

    public Node items(Node... items) {
        this.items = Arrays.asList(items);
        return this;
    }

    public Node properties(Map<String, Node> properties) {
        if (properties != null) {
            this.properties = new HashMap<>(properties);
        } else {
            this.properties = null;
        }
        return this;
    }

    public Node properties(String key1, Node value1) {
        if (this.properties == null) {
            this.properties = new HashMap<>();
        }
        this.properties.put(key1, value1);
        return this;
    }

    public Node properties(String key1, Node value1, String key2, Node value2) {
        properties(key1, value1);
        properties(key2, value2);
        return this;
    }

    public Node properties(String key1, Node value1, String key2, Node value2, String key3, Node value3) {
        properties(key1, value1, key2, value2);
        properties(key3, value3);
        return this;
    }

    public Node properties(String key1, Node value1, String key2, Node value2, String key3, Node value3, String key4, Node value4) {
        properties(key1, value1, key2, value2, key3, value3);
        properties(key4, value4);
        return this;
    }

    public Node blueId(String blueId) {
        this.blueId = blueId;
        return this;
    }

    public Node referenceBlueId(String referenceBlueId) {
        this.blueId = referenceBlueId;
        return this;
    }

    public Node computedBlueId(String computedBlueId) {
        this.computedBlueId = computedBlueId;
        return this;
    }

    public Node constraints(Constraints constraints) {
        this.constraints = constraints;
        return this;
    }

    public Node blue(Node blue) {
        this.blue = blue;
        return this;
    }
    
    public Node inlineValue(boolean inlineValue) {
        this.inlineValue = inlineValue;
        return this;
    }

    public Object get(String path) {
        return NodePathAccessor.get(this, path);
    }

    public Object get(String path, Function<Node, Node> linkingProvider) {
        return NodePathAccessor.get(this, path, linkingProvider);
    }

    public Node getAsNode(String path) {
        Object value = get(path);
        if (value instanceof Node) {
            return (Node) value;
        }
        throw new IllegalArgumentException("Value at path " + path + " is not a node: " + value);
    }

    public String getAsText(String path) {
        Object value = get(path);
        if (value == null) {
            return null;
        }
        if (value instanceof String) {
            return (String) value;
        }
        throw new IllegalArgumentException("Value at path " + path + " is not text: " + value);
    }

    public Integer getAsInteger(String path) {
        Object value = get(path);
        if (value instanceof BigInteger) {
            try {
                return ((BigInteger) value).intValueExact();
            } catch (ArithmeticException ex) {
                throw new IllegalArgumentException("Value at path " + path + " is out of Integer range: " + value, ex);
            }
        } else if (value instanceof BigDecimal) {
            BigDecimal bdValue = (BigDecimal) value;
            BigDecimal integralCandidate = bdValue.stripTrailingZeros();
            if (integralCandidate.scale() <= 0) {
                try {
                    return integralCandidate.intValueExact();
                } catch (ArithmeticException ex) {
                    throw new IllegalArgumentException("Value at path " + path + " is out of Integer range: " + value, ex);
                }
            } else {
                throw new IllegalArgumentException("Value at path " + path + " is not an integer: " + bdValue);
            }
        } else {
            throw new IllegalArgumentException("Value at path " + path + " is not a BigInteger or BigDecimal: " + value);
        }
    }

    @Override
    public Node clone() {
        try {
            Node cloned = (Node) super.clone();

            cloned.name = this.name;
            cloned.description = this.description;
            cloned.value = this.value;
            cloned.computedBlueId = this.computedBlueId;

            if (this.type != null) {
                cloned.type = this.type.clone();
            }

            if (this.itemType != null) {
                cloned.itemType = this.itemType.clone();
            }

            if (this.keyType != null) {
                cloned.keyType = this.keyType.clone();
            }

            if (this.valueType != null) {
                cloned.valueType = this.valueType.clone();
            }

            if (this.items != null) {
                cloned.items = this.items.stream()
                        .map(Node::clone)
                        .collect(Collectors.toCollection(ArrayList::new));
            }

            if (this.properties != null) {
                cloned.properties = this.properties.entrySet().stream()
                        .collect(Collectors.toMap(
                                Map.Entry::getKey,
                                entry -> entry.getValue().clone(),
                                (e1, e2) -> e1,
                                HashMap::new
                        ));
            }

            if (this.constraints != null) {
                cloned.constraints = this.constraints.clone();
            }

            if (this.blue != null) {
                cloned.blue = this.blue.clone();
            }

            return cloned;
        } catch (CloneNotSupportedException e) {
            throw new AssertionError("Node must be cloneable", e);
        }
    }

    @Override
    public String toString() {
        return "Node{" +
               "name='" + name + '\'' +
               ", description='" + description + '\'' +
               ", type=" + type +
               ", itemType=" + itemType +
               ", keyType=" + keyType +
               ", valueType=" + valueType +
               ", value=" + value +
               ", items=" + items +
               ", properties=" + properties +
               ", blueId='" + blueId + '\'' +
               ", computedBlueId='" + computedBlueId + '\'' +
               ", constraints=" + constraints +
               ", blue=" + blue +
               ", inlineValue=" + inlineValue +
               '}';
    }
}